### ISSUE TITLE
package-config-file-v2 spec

### ISSUE THREAD

**Issue by jensjoha** [dart-lang/language#772]
In the specification of the new package config (https://github.com/dart-lang/language/blob/master/accepted/future-releases/language-versioning/package-config-file-v2.md) it says

>Tools currently allow you to specify the .packages file with a command line argument:
>
>--packages=someFile
>-p someFile
[...]
> Tools should look for a .dart_tool/package_config.json relative to the specified file if that file turns out to be a .packages file.

I think that's a really bad idea.
If I tell for instance the compiler "use /path/to/my/.packages" I expect it to do so, I do not expect it to  arbitrarily (and behind my back) use some other file.

**Comment by lrhn** [comment]
The `.packages` file is intended as a temporary backwards compatibility help for third-party scripts.

The file still exists because there is existing third-party code out there *reading* the file, and we want that code to keep working. 
The Dart platform tools, however, should *never* read `.packages` again if there is a better file available. The `.packages` file does not contain sufficient information to make programs work. This won't be an issue until we release NNBD, but at that point, it will kick in.

If a third-party script currently runs `dart -p something/.packages script.dart`, and it would keep doing that after we get a language version supporting  NNBD, then the script may break.

If the script is not NNBD migrated, and its code will not run as NNBD code, then it *needs* language versioning to keep the library at a pre-NNBD language version. This information is available in `package_config.json`, but not in `.packages`. So, if running only with `.packages`, there will be no language version information for `script.dart`, and it will be run at the most recent language version, and it will fail. If we instead find the `package_config.json` file corresponding to the suggested `.packages` file, then the script code will be run as *it* intended, even though the third-party script running it didn't know that.

In short: Dart tools should just not read `.packages` if they can avoid it.
Having it automatically check for a `.dart_tool/package_config.json` file and fall back on the `.packages` file is a temporary measure to keep older code working.
And by looking for a `.dart_tool/package_config.json` *even* if a third-party tool asks the new Dart tool to look at `.packages` is a way to support that third-party tool for a little longer.

Long term, there will be no `.packages` file.

**Comment by jensjoha** [comment]
The way I see it there's drawbacks in every scenario:
* You specify .packages and you get what you say: Your point --- you won't get language versioning. May break third-party scripts in the future, but at least the tool does what it's asked.
* You specify .packages and get .json: Your point --- you get language versioning and it *might* be what the user wanted. Saying that it's "for your own good" and a "better file" is hand-wavy at best though, and still goes against what the user asked (imagine ordering something at a restaurant and getting something else because the waiter "knows better"). There's two variations of this:
  * Do it silently: The user has no idea what's going on and we can get into situations where the user spends hours and hours trying to figure out why the compiler doesn't do what it's been asked.
  * Emit a warning: When called manually at least the user knows, but may break scripts that checks that the output from a compile is non-empty, or uses the whole output as the result of a calculation or whatnot.

_My_ conclusion still is that there is no _good_ solution, and that the _best_ one is you do what the user asks.

Ultimately, of cause, this is not my decision --- but at least the spec should be very clear, both about which file it picks and whether a warning is issued. While I may be reading it wrong, "should" sounds like a suggestion to me (https://tools.ietf.org/html/rfc2119 basically equates 'should' with 'recommend' and https://www.merriam-webster.com/dictionary/shall specifies synonyms as both 'have (to)' and 'ought (to)' again specified as "must" and (again among other things) "advisability").

**Comment by lrhn** [comment]
The tool guideline is to *only* use `package_config.json`. The `.packages` file is deprecated and will be removed eventually. Any use of the .packages` file  is there only for backwards compatibility with tools which doesn't know to use `package_config.json` yet yet. 
Specifically asking for `.packages` proves that the source is such a tool, so in that case, the waited actually does know better.

I agree that we should decide whether this is mandatory and update the phrasing to match.

**Comment by jensjoha** [comment]
An update on this is that I was told - offline - that the tool _has_ to look for the json file, so what's what happened.
